code/modules/research/ordnance/tank_compressor.dm

```diff
@@ -0,0 +1,330 @@
+	var/record_number = 1
+	var/obj/item/tank/inserted_tank
+	/// Reference to a disk we are going to print to.
+	var/obj/item/computer_hardware/hard_drive/portable/inserted_disk
+
+	pipe_flags = PIPING_ONE_PER_TURF | PIPING_DEFAULT_LAYER_ONLY
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/Initialize(mapload)
+	. = ..()
+	leaked_gas_buffer = new(200)
+	compressor_record = list()
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/ComponentInitialize()
+	. = ..()
+	RegisterSignal(src, COMSIG_ATOM_INTERNAL_EXPLOSION, .proc/explosion_handle)
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/examine()
+	. = ..()
+	. += "This one is rated for up to [TANK_COMPRESSOR_PRESSURE_LIMIT] kPa."
+
+/// Stores the record of the gas data for a significant enough tank leak
+/datum/data/compressor_record
+	/// Tank Name
+	var/experiment_source
+	/// Key: Path, Value: Moles
+	var/list/gas_data = list()
+	var/timestamp
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/attackby(obj/item/item, mob/living/user)
+	if (!panel_open)
+		if(istype(item, /obj/item/tank))
+			var/obj/item/tank/tank_item = item
+			if(inserted_tank)
+				if(!eject_tank(user))
+					balloon_alert(user, span_warning("[inserted_tank] is stuck inside."))
+					return ..()
+			if(!user.transferItemToLoc(tank_item, src))
+				balloon_alert(user, span_warning("[tank_item] is stuck to your hand."))
+				return ..()
+			inserted_tank = tank_item
+			last_recorded_pressure = 0
+			RegisterSignal(inserted_tank, COMSIG_PARENT_QDELETING, .proc/unregister_tank)
+			update_appearance()
+			return
+		if(istype(item, /obj/item/computer_hardware/hard_drive/portable))
+			var/obj/item/computer_hardware/hard_drive/portable/attacking_disk = item
+			eject_disk(user)
+			if(user.transferItemToLoc(attacking_disk, src))
+				inserted_disk = attacking_disk
+			else
+				balloon_alert(user, span_warning("[attacking_disk] is stuck to your hand."))
+			return
+	if(!active && !inserted_tank)
+		if(default_deconstruction_screwdriver(user, "[base_icon_state]-open", "[base_icon_state]-open", item))
+			update_appearance()
+			return
+		if(panel_open)
+			if(default_deconstruction_crowbar(item))
+				return
+			if(default_change_direction_wrench(user, item))
+				return
+	return ..()
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/default_change_direction_wrench(mob/user, obj/item/wrench)
+	. = ..()
+	if(.)
+		/// Disconnect our partner.
+		if(nodes[1])
+			nodes[1].disconnect(src)
+			nodes[1] = null
+			if(parents[1])
+				nullify_pipenet(parents[1])
+		if(nodes[2])
+			nodes[2].disconnect(src)
+			nodes[2] = null
+			if(parents[2])
+				nullify_pipenet(parents[2])
+		set_init_directions()
+		/// Connect to a new one.
+		atmos_init()
+		if(nodes[1])
+			nodes[1].atmos_init()
+			nodes[1].add_member(src)
+		if(nodes[2])
+			nodes[2].atmos_init()
+			nodes[2].add_member(src)
+		SSair.add_to_rebuild_queue(src)
+		update_appearance()
```

Comment:
God we need a helper proc for this someday

Reply:
holding out for a hero